### PR TITLE
feat: osmosis spot prices ticker

### PR DESCRIPTION
**Pull request by adamewozniak**
## Description

This PR implements the initial ticker object for querying prices from osmosis.

Just filling with spot price initially, volume to come - that will require some msg querying until osmosis's [tvwap queries](https://github.com/osmosis-labs/osmosis/tree/main/x/gamm/twap) are available.

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: sync/sync.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package sync
```
Comment (adamewozniak) [inline sync/sync.go:1]: This package comes from the price feeder, but the version of the SDK on PF is incompatible with osmosis's request lib

Comment (RafilxTenfen) [reply sync/sync.go:1]: Can't we import the Closer from price-feeder? the only package that it imports is "sync"

File: api/types.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package api
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// Ticker serves to hold the pricing data for a given pool.
+type Ticker struct {
```
Comment (RafilxTenfen) [inline api/types.go:10]: Maybe use the [TickerPrice](https://github.com/umee-network/umee/blob/58be905c40d42791cc4f2dac4e99f865a1832ddb/price-feeder/oracle/provider/provider.go#L39-L40)?

File: api/api.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+			Price: ticker,
+		}
+	}
+
+	return nil
+}
+
+// GetTickers returns a new instance of the current pool tickers.
+func (api *PriceAPI) GetTickers() map[string]Ticker {
+	api.mtx.Lock()
```
Comment (RafilxTenfen) [inline api/api.go:94]:
```suggestion
	api.mtx.RLock()
```

File: api/api.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		logger:  logger.With().Str("module", "price_api").Logger(),
+		closer:  osync.NewCloser(),
+		client:  client,
+		tickers: make(map[string]Ticker),
+		pools:   pools,
+	}
+}
+
+// Start begins the API process in a blocking fashion.
+func (a *PriceAPI) Start(ctx context.Context) error {
```
Comment (RafilxTenfen) [inline api/api.go:49]:
```suggestion
func (a *PriceAPI) Start(ctx context.Context) {
```

File: api/api.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+			return err
+		}
+
+		api.tickers[pool.String()] = Ticker{
+			Price: ticker,
+		}
+	}
+
+	return nil
+}
```
Comment (RafilxTenfen) [inline api/api.go:90]:
```suggestion
// SetTickers queries the osmosis node and sets the current price.
func (api *PriceAPI) SetTickers(ctx context.Context) error {
        api.mtx.Lock()
        defer api.mtx.Unlock()
        
	for _, pool := range api.pools {
		price, err := api.client.GetSpotPrice(ctx, types.QuerySpotPriceRequest{
			PoolId:        pool.PoolId,
			TokenInDenom:  pool.Base,
			TokenOutDenom: pool.Quote,
		})
		if err != nil {
			return err
		}

		ticker, err := sdk.NewDecFromStr(price.SpotPrice)
		if err != nil {
			return err
		}

		api.tickers[pool.String()] = Ticker{
			Price: ticker,
		}
	}

	return nil
}
```

**Review by RafilxTenfen — CHANGES_REQUESTED** [review]

File: api/api.go (hunk 5 above)
Comment (RafilxTenfen) [reply api/api.go:90]: - It was not intended to modify all of this, just added Lock/Unlock at the beginning

File: api/api.go (hunk 3 above)
Comment (adamewozniak) [reply api/api.go:94]: Thank you for this find 🙏

File: sync/sync.go (hunk 1 above)
Comment (adamewozniak) [reply sync/sync.go:1]: @RafilxTenfen we upgraded to osmosis v7 in #23 and it fixed the issue, I'll use that package here after that gets merged in

**Review by RafilxTenfen — APPROVED** [review]

File: api/api.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+			a.closer.Close()
+
+		default:
+			a.logger.Debug().Msg("starting api tick")
+
+			if err := a.tick(ctx); err != nil {
+				a.logger.Err(err).Msg("api tick failed")
+			}
+
+			time.Sleep(tickerSleep)
```
Comment (robert-zaremba) [inline api/api.go:62]: we are mixing globals with a struct context. `tickerSleep` should be part of the PriceAPI structure

**Pull request merged**